### PR TITLE
 BUG: RecursionError when attempting to replace np.nan values (#45725)

### LINKED ISSUE #45725: BUG: RecursionError when attempting to replace `np.nan` values under main branch
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np


NaN_values = [np.nan, np.nan]
ser = pd.Series(data=NaN_values)
df = pd.DataFrame(data=NaN_values)


try:
    ser = ser.replace({np.nan: pd.NA})
except RecursionError:
    print("RecursionError: maximum recursion depth exceeded while calling a Python object")

try:
    df = df.replace({np.nan: None})
except RecursionError:
    print("RecursionError: maximum recursion depth exceeded while calling a Python object")
```

### Issue Description

Note: this bug is not present in 1.4.x branch, but it is in the main branch and [1.5 milestone](https://github.com/pandas-dev/pandas/milestone/92).

Let's say we have a Series or DataFrame containing `numpy.nan` values (a common scenario), and we want to convert them all to Panda's own type `pandas.NA`, or alternatively to the native Python `None` type. One way to do this conversion is by using the `replace` method as shown in the example. From my testing this works in pandas 1.4.0 as well as 1.3.5, but not in the main branch. A RecursionError exception is thrown.

Suggested tags:  `1.5 Milestone`, `Regression`, `replace`, `Dtype Conversions`

### Expected Behavior

The `replace` method should successfully replace `np.nan` with the desired value, without errors. This behavior works in pandas 1.4.0 and previous versions and should be fixed for the pandas 1.5 release.

### Installed Versions

<details>
commit           : f77480f9f5ed068d7e726862dff2db0411015201

python           : 3.10.2.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.22000
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.0.dev0+234.gf77480f9f5
numpy            : 1.22.1
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.1.2
setuptools       : 57.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by roib20**
- [x] closes #45725
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.5.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by roib20** [comment]
@github-actions pre-commit

File: pandas/tests/frame/methods/test_replace.py [hunk 1]
```diff
@@ -661,6 +661,18 @@ def test_replace_simple_nested_dict_with_nonexistent_value(self):
...
         tm.assert_frame_equal(expected, result)
 
+    def test_replace_numpy_nan(self):
+        # GH#45725 ensure np.nan can be replaced with pd.NA
+        df = pd.DataFrame({"A": [np.nan], "B": [np.nan]}, dtype=object)
+        result = df.replace({np.nan: pd.NA})
+        expected = pd.DataFrame({"A": [pd.NA], "B": [pd.NA]}, dtype=object)
+        tm.assert_frame_equal(result, expected)
+
+        # same thing but with None
```
Comment (mroeschke) [inline pandas/tests/frame/methods/test_replace.py:671]: Could you use `pytest.mark.parameterize` with `pd.NA/None` (same for the other test)?

**Comment by roib20** [comment]
@github-actions pre-commit

File: pandas/tests/frame/methods/test_replace.py (hunk 1 above)
Comment (roib20) [reply pandas/tests/frame/methods/test_replace.py]: Done.

File: pandas/tests/frame/methods/test_replace.py [hunk 2]
```diff
@@ -661,6 +661,15 @@ def test_replace_simple_nested_dict_with_nonexistent_value(self):
         result = df.replace({"col": {-1: "-", 1: "a", 4: "b"}})
         tm.assert_frame_equal(expected, result)
 
+    @pytest.mark.parametrize("dtype", [object])
```
Comment (mroeschke) [inline pandas/tests/frame/methods/test_replace.py:664]: This only has one param so it can be in-lined (same as below)

Comment (roib20) [reply pandas/tests/frame/methods/test_replace.py:664]: Alright done. Thank you for the super-fast response.

**Review by mroeschke — APPROVED** [review]

File: doc/source/whatsnew/v1.5.0.rst [hunk 3]
```diff
@@ -248,6 +248,7 @@ Conversion
 - Bug in :meth:`Series.astype` and :meth:`DataFrame.astype` from floating dtype to unsigned integer dtype failing to raise in the presence of negative values (:issue:`45151`)
 - Bug in :func:`array` with ``FloatingDtype`` and values containing float-castable strings incorrectly raising (:issue:`45424`)
 - Bug when comparing string and datetime64ns objects causing ``OverflowError`` exception. (:issue:`45506`)
+- Bug when attempting to replace ``numpy.nan`` values causing ``RecursionError``
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:251]: if i read the original issue correctly this then this was working in 1.4 so no actual change (its only a change in master), if so delete this note

File: pandas/tests/frame/methods/test_replace.py [hunk 4]
```diff
@@ -661,6 +661,14 @@ def test_replace_simple_nested_dict_with_nonexistent_value(self):
         result = df.replace({"col": {-1: "-", 1: "a", 4: "b"}})
         tm.assert_frame_equal(expected, result)
 
+    @pytest.mark.parametrize("to_replace, value", [(np.nan, pd.NA), (np.nan, None)])
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:664]: can you also add `np.nan` as a target

File: pandas/tests/frame/methods/test_replace.py [hunk 5]
```diff
@@ -661,6 +661,14 @@ def test_replace_simple_nested_dict_with_nonexistent_value(self):
...
+    def test_replace_numpy_nan(self, to_replace, value):
+        # GH#45725 ensure numpy.nan can be replaced with pandas.NA or None
+        df = DataFrame({"A": [to_replace]}, dtype=object)
+        result = df.replace({to_replace: value})
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:668]: can you also test usage with `df.replace(to_replace, value)` or is this tested close by here?

File: pandas/tests/series/methods/test_replace.py [hunk 6]
```diff
@@ -36,6 +36,15 @@ def test_replace_explicit_none(self):
         assert expected.iloc[-1] is None
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize("to_replace, value", [(np.nan, pd.NA), (np.nan, None)])
```
Comment (jreback) [inline pandas/tests/series/methods/test_replace.py:39]: same comments as above. consider just using the null_fixture which has all the nans

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.5.0.rst (hunk 3 above)
Comment (roib20) [reply doc/source/whatsnew/v1.5.0.rst]: Yes you're right. I removed the note from the changelog.

File: pandas/tests/series/methods/test_replace.py (hunk 6 above)
Comment (roib20) [reply pandas/tests/series/methods/test_replace.py]: `nulls_fixture` was a good solution. Implemented.

File: pandas/tests/frame/methods/test_replace.py (hunk 5 above)
Comment (roib20) [reply pandas/tests/frame/methods/test_replace.py]: Yes done.

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @roib20 very nice! keep em coming!

**Comment by simonjayhawkins** [comment]
The tests added in this PR don't seem to fail on master with `value = self._standardize_fill_value(value) # GH#45725` removed.

Also before #45568, which maybe the cause of the regression, the values were replaced.

```
2022-03-16T22:14:48.3750817Z 1.5.0.dev0+209.g0f5d934be7
2022-03-16T22:14:48.3751119Z 0    <NA>
2022-03-16T22:14:48.3751330Z 1    <NA>
2022-03-16T22:14:48.3751555Z dtype: object
2022-03-16T22:14:48.3751789Z       0
2022-03-16T22:14:48.3751988Z 0  None
2022-03-16T22:14:48.3752170Z 1  None
```

from bisect with following code sample

```
# BUG: RecursionError when attempting to replace np.nan values under main branch #45725


import numpy as np

import pandas as pd

print(pd.__version__)

NaN_values = [np.nan, np.nan]
ser = pd.Series(data=NaN_values)
df = pd.DataFrame(data=NaN_values)


ser = ser.replace({np.nan: pd.NA})
print(ser)

df = df.replace({np.nan: None})
print(df)
```

on master, same code sample gives

```
1.5.0.dev0+544.g49937b935d
0   NaN
1   NaN
dtype: float64
    0
0 NaN
1 NaN
```

not sure yet, but I think we will need to remove the ``value = self._standardize_fill_value(value) # GH#45725` added in this PR to fix #45601 and #45836

might propose reverting this PR until the regressions in 1.4.0 are fixed

**Comment by simonjayhawkins** [comment]
> The tests added in this PR don't seem to fail on master with `value = self._standardize_fill_value(value) # GH#45725` removed.

but to be clear, the recursion error still exists when the fix added here is removed. Looks like need to revisit these tests.

For the record the traceback is currently

```python-traceback
Traceback (most recent call last):
  File "/home/simon/45725.py", line 15, in <module>
    ser = ser.replace({np.nan: pd.NA})
  File "/home/simon/pandas/pandas/core/series.py", line 5000, in replace
    return super().replace(
  File "/home/simon/pandas/pandas/core/generic.py", line 6681, in replace
    return self.replace(
  File "/home/simon/pandas/pandas/core/series.py", line 5000, in replace
    return super().replace(
  File "/home/simon/pandas/pandas/core/generic.py", line 6730, in replace
    new_data = self._mgr.replace_list(
  File "/home/simon/pandas/pandas/core/internals/managers.py", line 421, in replace_list
    bm = self.apply(
  File "/home/simon/pandas/pandas/core/internals/managers.py", line 304, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 725, in replace_list
    result = blk._replace_coerce(
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 780, in _replace_coerce
    return self.replace(
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 605, in replace
    return blk.replace(
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 605, in replace
    return blk.replace(
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 605, in replace
    return blk.replace(
  [Previous line repeated 977 more times]
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 604, in replace
    blk = self.coerce_to_target_dtype(value)
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 449, in coerce_to_target_dtype
    return self.astype(new_dtype, copy=False)
  File "/home/simon/pandas/pandas/core/internals/blocks.py", line 523, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
  File "/home/simon/pandas/pandas/core/dtypes/astype.py", line 283, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
  File "/home/simon/pandas/pandas/core/dtypes/astype.py", line 215, in astype_array
    if is_datetime64tz_dtype(dtype) and is_datetime64_dtype(values.dtype):
  File "/home/simon/pandas/pandas/core/dtypes/common.py", line 386, in is_datetime64tz_dtype
    return DatetimeTZDtype.is_dtype(arr_or_dtype)
  File "/home/simon/pandas/pandas/core/dtypes/base.py", line 312, in is_dtype
    if isinstance(dtype, (ABCSeries, ABCIndex, ABCDataFrame, np.dtype)):
  File "/home/simon/pandas/pandas/core/dtypes/generic.py", line 45, in _check
    return getattr(inst, attr, "_typ") in comp
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by simonjayhawkins** [comment]
Because the behavior no longer matches the pre-regression behavior and the fact that the tests do not actually catch the reported regression, I now propose to revert this PR and reopen #45725.

Even though this fix prevents the regression, I think it will be cleaner to undo this and redo properly rather than patching it up.

**Pull request merged**